zerver/views/registration.py

```diff
@@ -247,6 +247,11 @@ def accounts_register(request: HttpRequest) -> HttpResponse:
                                        return_data=return_data)
             if auth_result is not None:
                 # Since we'll have created a user, we now just log them in.
+                process_prereg_user(auth_result, prereg_user, default_stream_groups)
+                if is_guest:
+                    do_change_is_guest(auth_result, is_guest)
+                elif is_realm_admin:
+                    do_change_is_admin(auth_result, is_realm_admin)
```

Comment:
This part I don't understand -- I think instead we should pass the `prereg_user` to `authenticate` so that we can use it in `do_create_user` there.

Reply:
(I think what you have will result in calling some functions twice)

Reply:
My reasoning for choosing this approach was that `process_prereg_user()` is an idempotent function so even if we call it twice it shouldn't cause any problem. But your approach definitely seems better. :)